src/ui/public/chrome/directives/app_switcher/app_switcher.js

```diff
@@ -6,14 +6,45 @@ var DomLocationProvider = require('ui/domLocation');
 
 require('ui/modules')
 .get('kibana')
+.provider('appSwitcherEnsureNavigation', function () {
+  var forceNavigation = false;
+
+  this.forceNavigation = function (val) {
+    forceNavigation = !!val;
+  };
+
+  this.$get = ['Private', function (Private) {
+    var domLocation = Private(DomLocationProvider);
+
+    return function (event) {
+      if (!forceNavigation || event.isDefaultPrevented() || event.altKey || event.metaKey || event.ctrlKey) {
+        return;
+      }
+
+      var toParsed = parse(event.delegateTarget.href);
+      var fromParsed = parse(domLocation.href);
+      var sameProto = toParsed.protocol === fromParsed.protocol;
+      var sameHost = toParsed.host === fromParsed.host;
+      var samePath = toParsed.path === fromParsed.path;
+
+      if (sameProto && sameHost && samePath) {
+        toParsed.hash && domLocation.reload();
+
+        // event.preventDefault() keeps the browser from seeing the new url as an update
+        // and even setting window.location does not mimic that behavior, so instead
+        // we use stopPropagation() to prevent angular from seeing the click and
+        // starting a digest cycle/attempting to handle it in the router.
+        event.stopPropagation();
+      }
+    };
+  }];
+})
 .directive('appSwitcher', function () {
   return {
     restrict: 'E',
     template: require('./app_switcher.html'),
     controllerAs: 'switcher',
-    controller: function ($scope, Private) {
-      var domLocation = Private(DomLocationProvider);
-
+    controller: function ($scope, appSwitcherEnsureNavigation) {
```

Comment:
If you use the short function syntax from es6 here, then the resulting function will be properly named for better stack traces:
```
controller($scope, appSwitcherEnsureNavigation) {
},
```